crates/transaction-pool/src/pool/pending.rs

```diff
@@ -325,10 +347,63 @@ impl<T: TransactionOrdering> PendingPool<T> {
         id
     }
 
-    /// Removes the worst transaction from this pool.
-    pub(crate) fn pop_worst(&mut self) -> Option<Arc<ValidPoolTransaction<T::Transaction>>> {
-        let worst = self.all.iter().next().map(|tx| *tx.transaction.id())?;
-        self.remove_transaction(&worst)
+    /// Truncates the pool to the given limit.
+    ///
+    /// Removes transactions from the pending pool
+    /// The algorithm tries to reduce transaction counts by an approximately
+    /// equal number for all for accounts with many pending transactions.
+    pub(crate) fn truncate_pool(
+        &mut self,
+        limit: SubPoolLimit,
+    ) -> Vec<Arc<ValidPoolTransaction<T::Transaction>>> {
+        // return early if nothing needs to happen
+        if self.size() <= limit.max_size && self.len() <= limit.max_txs {
+            return Vec::new()
+        }
+
+        let mut removed = Vec::new();
+
+        // penalize non-local txs if limit is still exceeded
+        while self.size() > limit.max_size || self.len() > limit.max_txs {
+            // flag to break out of the loop if all are local
+            let mut any_non_local = true;
+
+            // use descendants, so we pop a single tx per sender
+            for tx in self.highest_nonces.clone().iter() {
```

Comment:
even though this only includes unique senders and we expect that the outer loop is really a single iterator.
I think we get rid of the clone if we instead push all txs ids that need to be removed to a tmp vec, keep track of the amount and sizes and only remove them once we have collected enough, then do the same again without locals check if necessary

Reply:
tried to implement this with `transactions_to_remove`